src/shared/dissect-image.c

```diff
@@ -810,21 +813,22 @@ static int dissect_image(
                 generic_rw = false,  /* initialize to appease gcc */
                 generic_growfs = false;
         _cleanup_(blkid_free_probep) blkid_probe b = NULL;
-        _cleanup_free_ char *generic_node = NULL;
+        _cleanup_free_ char *generic_node = NULL, *root_fstype_string = NULL;
         sd_id128_t generic_uuid = SD_ID128_NULL;
-        const char *pttype = NULL;
+        const char *pttype = NULL, *root_fs_usage = NULL;
         blkid_partlist pl;
+        PartitionPolicyFlags root_fstype = 0;
         int r, generic_nr = -1, n_partitions;
 
         assert(m);
-        assert(fd >= 0);
+        assert(fd >= 0 || (flags & DISSECT_IMAGE_NO_PARTITION_TABLE));
```

Comment:
i *really* dislike having a function called "dissect" that doesn't dissect.

please just add a new helper dissect_image_from_policy() or so which tries to fill in a DissectedImage structure from the policy alone, and fails with EAGAIN or so if the policy is not complete enough for that.

and if you don#t want to duplicate parts of the existing dissect_image() func for that, then split out the common part into new helpers that you call from both